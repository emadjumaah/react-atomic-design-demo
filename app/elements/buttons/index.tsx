import React, { FunctionComponent } from 'react';
import { IDefaultProps } from 'types';
import styled from 'styled-components';

export enum ButtonVariant {
  PRIMARY = 'primary',
  SECONDARY = 'secondary',
  TERTIARY = 'tertiary',
}

export enum ButtonColor {
  BLUE = 'primary',
  ORANGE = 'secondary',
  GREEN = 'tertiary',
  WHITE = 'white',
  BLACK = 'black',
}

interface IButtonProps extends IDefaultProps {
  /**
   * [required] Children must be a string
   */
  children: string;
  /**
   * [required] Callback to trigger when user clicks on button
   */
  onClick: () => void;
  /**
   * [optional] Color of button
   *
   * @default ButtonColor.BLUE
   */
  color?: ButtonColor;
}

interface IBaseButtonProps extends IButtonProps {
  /**
   * [optional] Variant of button to use
   *
   * @default ButtonVariant.PRIMARY
   */
  variant?: ButtonVariant;
}

const BaseButton: FunctionComponent<IBaseButtonProps> = props => (
  <button data-testid="button" className={props.className} onClick={props.onClick}>
    {props.children}
  </button>
);

/**
 * @todo - Refactor how colors are used within application to support better accessibility
 * and possibly the ability to use tools like Polished's `lighten` or `darken` for better
 * hover & press effects with buttons
 */
const getBackgroundColor = (variant: ButtonVariant, color: ButtonColor) => {
  switch (variant) {
    case ButtonVariant.PRIMARY:
      return `
          background-color: var(--${color});
        `;
    case ButtonVariant.SECONDARY:
    case ButtonVariant.TERTIARY:
    default:
      return `
          background-color: transparent;
        `;
  }
};

const getBorderColor = (variant: ButtonVariant, color: ButtonColor) => {
  switch (variant) {
    case ButtonVariant.SECONDARY:
      return `
        border: 2px solid var(--${color});
      `;
    case ButtonVariant.PRIMARY:
    case ButtonVariant.TERTIARY:
    default:
      return `
        border: none
      `;
  }
};

const getFontColor = (variant: ButtonVariant, color: ButtonColor) => {
  switch (variant) {
    case ButtonVariant.PRIMARY:
      return color === ButtonColor.WHITE
        ? `color: var(--black);`
        : `color: var(--white);`;
    case ButtonVariant.SECONDARY:
    case ButtonVariant.TERTIARY:
    default:
      return `
        color: var(--${color});
      `;
  }
};

const getButtonCss = (props: IBaseButtonProps) => {
  const color = props.color || ButtonColor.BLUE;
  const variant = props.variant || ButtonVariant.PRIMARY;

  return `
    ${getBackgroundColor(variant, color)}
    ${getBorderColor(variant, color)}
    ${getFontColor(variant, color)}

    &:hover {
      cursor: pointer;
    }
  `;
};

const StyledBaseButton = styled(BaseButton)`
  ${props => getButtonCss(props)}
`;

export const PrimaryButton: FunctionComponent<IButtonProps> = props => (
  <StyledBaseButton {...props} variant={ButtonVariant.PRIMARY} />
);
export const SecondaryButton: FunctionComponent<IButtonProps> = props => (
  <StyledBaseButton {...props} variant={ButtonVariant.SECONDARY} />
);
export const TertiaryButton: FunctionComponent<IButtonProps> = props => (
  <StyledBaseButton {...props} variant={ButtonVariant.TERTIARY} />
);
