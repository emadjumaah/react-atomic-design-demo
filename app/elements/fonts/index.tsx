import React from 'react';
import { IDefaultProps } from 'types';

export const PageTitle = (props: IDefaultProps) => <h1 {...props} />;
export const SectionTitle = (props: IDefaultProps) => <h2 {...props} />;
export const SubsectionTitle = (props: IDefaultProps) => <h3 {...props} />;

interface IParagraphFontProps extends IDefaultProps {
  /**
   * [optional] Wraps children in `em` tag
   */
  emphasis?: boolean;
  /**
   * [optional] Wraps children in `strong` tag
   */
  strong?: boolean;
  /**
   * [optiona] Renders paragraph with Description instead of Primary styles
   */
  isDescription?: boolean;
}

const Emphasis = (props: any) => <em {...props} />;
const Strong = (props: any) => <strong {...props} />;
export const PrimaryText = (props: IParagraphFontProps) => {
  const className = props.isDescription
    ? `${props.className} description`
    : props.className;
  if (props.emphasis || props.strong) {
    const Em = props.emphasis ? Emphasis : React.Fragment;
    const St = props.strong ? Strong : React.Fragment;
    return (
      <p className={className}>
        <Em>
          <St>{props.children}</St>
        </Em>
      </p>
    );
  }
  return <p {...props} className={className} />;
};
export const DescriptionText = (props: IParagraphFontProps) => (
  <PrimaryText {...props} className={`${props.className} description`} />
);
