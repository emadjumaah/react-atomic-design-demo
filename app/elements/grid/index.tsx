import React, { FunctionComponent } from 'react';
import { IDefaultProps } from 'types';
import { Row } from 'layouts';

export type TRowColumn = [number, number];

export interface IGridProps extends IDefaultProps {
  /**
   * [required] Grid size [rows, columns]
   */
  grid: TRowColumn;
  /**
   * [required] Callback responsible for returning a valid JSX Element based
   * on current row/column position
   */
  renderCell: (row: number, column: number) => JSX.Element;
}

export const Grid: FunctionComponent<IGridProps> = props => {
  const rows = new Array(props.grid[0]).fill(0);
  const columns = new Array(props.grid[1]).fill(0);
  return (
    <>
      {rows.map((_, rIndex) => (
        <Row key={`row-${rIndex}`}>
          {columns.map((_, cIndex) => props.renderCell(rIndex, cIndex))}
        </Row>
      ))}
    </>
  );
};
