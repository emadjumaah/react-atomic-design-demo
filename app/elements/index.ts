export * from './fonts';
export * from './markdown';
export * from './navigation';
export * from './grid';
export * from './shapes';
export * from './buttons';
export * from './forms';
