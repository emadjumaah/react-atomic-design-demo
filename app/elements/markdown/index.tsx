import React, { FunctionComponent } from 'react';
import ReactMarkdown from 'react-markdown';
import { Column } from 'layouts';
import { markdownRenderers } from './components';

interface IMarkdownProps {
  /**
   * [required] Content source to convert from markdown to React components
   */
  source: string;
}

export const Markdown: FunctionComponent<IMarkdownProps> = props => (
  <Column>
    <ReactMarkdown {...props} renderers={markdownRenderers} />
  </Column>
);

export * from './components';
export * from './interactive-markdown';
