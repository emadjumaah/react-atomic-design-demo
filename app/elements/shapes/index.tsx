import React, { FunctionComponent } from 'react';
import styled from 'styled-components';
import { IDefaultProps } from 'types';
import { FlexSpace } from 'layouts';
import { Color } from 'config/colors';

interface IBaseShapeProps extends IDefaultProps {
  /**
   * [optional] Background color for shape
   *
   * @default Color.BLACK
   */
  backgroundColor?: Color;
}

const BaseShape: FunctionComponent<IBaseShapeProps> = props => <div {...props} />;

interface ICircleProps extends IBaseShapeProps {
  /**
   * [optional] Size of circle (diameter)
   *
   * @default FlexSpace.MEDIUM
   */
  size?: FlexSpace;
}

export const Circle = styled(BaseShape)<ICircleProps>`
  border-radius: 100%;
  height: ${props => props.size || FlexSpace.MEDIUM};
  width: ${props => props.size || FlexSpace.MEDIUM};
  margin: 8px;
  background-color: ${props => props.backgroundColor || Color.BLACK};
`;
