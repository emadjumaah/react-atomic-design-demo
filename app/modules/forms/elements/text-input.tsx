import React, { FunctionComponent, useContext, useEffect } from 'react';
import { ITextInputProps, TextInput as BaseTextInput } from 'elements';
import { FormContext, FormActions } from 'modules/forms';

interface IFormTextInput extends ITextInputProps {
  /**
   * [required] Unique identifier for text input within form
   */
  name: string;
}

export const TextInput: FunctionComponent<IFormTextInput> = ({ name, ...props }) => {
  const { dispatch } = useContext(FormContext);

  useEffect(() => {
    dispatch(FormActions.registerFormElement(name));
  }, []);

  return <BaseTextInput {...props} />;
};
