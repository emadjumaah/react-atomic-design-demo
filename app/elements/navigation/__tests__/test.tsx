import React from 'react';
import { render } from '@testing-library/react';
import { Link } from 'elements';

describe('Navigation', () => {
  it('Link Renders', () => {
    const linkUrl = 'http://www.link-url.com';
    const { container } = render(<Link href={linkUrl}>Navigation Link</Link>);
    expect(container).toMatchSnapshot();
    expect(container.querySelector('a').href).toContain(linkUrl);
  });
});
