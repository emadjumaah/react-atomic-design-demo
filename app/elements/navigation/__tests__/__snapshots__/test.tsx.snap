// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Link Renders 1`] = `
<div>
  <a
    href="http://www.link-url.com"
  >
    Navigation Link
  </a>
</div>
`;
